File: src/python/cable_joints/common_systems.py
import math
import numpy as np
from . import vector2 as v2_helpers
from .ecs import (
    PositionComponent,
    VelocityComponent,
    GravityAffectedComponent,
    MassComponent,
    PrevFinalPosComponent,
    OrientationComponent,
    MomentOfInertiaComponent,
    AngularVelocityComponent,
    DistanceConstraintComponent,
    PrevFinalOrientationComponent
)


class PrevFinalOrientationSystem:
    run_in_pause = False
    def update(self, world, dt):
        entities = world.query([OrientationComponent, PrevFinalOrientationComponent])
        for entity_id in entities:
            orientation_comp = world.get_component(entity_id, OrientationComponent)
            prev_final_orientation_comp = world.get_component(entity_id, PrevFinalOrientationComponent)
            prev_final_orientation_comp.angle = orientation_comp.angle

class PBDAngularVelocityUpdateSystem:
    run_in_pause = False

    def _normalize_angle(self, angle): # Helper to find the shortest angle difference
        while angle > math.pi:
            angle -= 2 * math.pi
        while angle < -math.pi:
            angle += 2 * math.pi
        return angle

    def update(self, world, dt):
        grabbed = world.get_resource('grabbedBall')
        entities = world.query([
            OrientationComponent,
            AngularVelocityComponent,
            PrevFinalOrientationComponent,
            MomentOfInertiaComponent
        ])
        epsilon = 1e-9

        if dt <= epsilon:
            return

        for entity_id in entities:
            if entity_id == grabbed:
                continue

            moi_comp = world.get_component(entity_id, MomentOfInertiaComponent)
            if moi_comp and moi_comp.inv_inertia <= 0: # Skip static or non-rotational objects
                continue

            orientation_comp = world.get_component(entity_id, OrientationComponent)
            angular_vel_comp = world.get_component(entity_id, AngularVelocityComponent)
            prev_final_orientation_comp = world.get_component(entity_id, PrevFinalOrientationComponent)

            current_angle = orientation_comp.angle
            prev_angle = prev_final_orientation_comp.angle

            delta_angle = self._normalize_angle(current_angle - prev_angle)

            angular_vel_comp.angular_velocity = delta_angle / dt

class PBDVelocityUpdateSystem:
    run_in_pause = False
    def update(self, world, dt):
        grabbed = world.get_resource('grabbedBall')
        entities = world.query([PositionComponent, VelocityComponent, PrevFinalPosComponent, MassComponent])
        epsilon = 1e-9

        if dt <= epsilon: # Avoid division by zero or very small dt
            return

        for entity_id in entities:
            if entity_id == grabbed: # Don't update velocity of grabbed ball this way
                continue

            mass_comp = world.get_component(entity_id, MassComponent)
            # Only update velocities for dynamic objects (mass > 0)
            if mass_comp and mass_comp.mass <= 0:
                continue

            pos_comp = world.get_component(entity_id, PositionComponent)
            vel_comp = world.get_component(entity_id, VelocityComponent)
            prev_final_pos_comp = world.get_component(entity_id, PrevFinalPosComponent)

            # v = (x_new - x_old) / dt
            np.subtract(pos_comp.pos, prev_final_pos_comp.pos, out=vel_comp.vel)
            np.divide(vel_comp.vel, dt, out=vel_comp.vel)


class GravitySystem:
    run_in_pause = False
    def update(self, world, dt):
        grabbed = world.get_resource('grabbedBall')
        gravity = world.get_resource('gravity')
        if gravity is None:
            return

        entities = world.query([VelocityComponent, GravityAffectedComponent])
        for entity_id in entities:
            if entity_id == grabbed:
                continue
            vel_comp = world.get_component(entity_id, VelocityComponent)
            vel_comp.vel += gravity * dt

class XPBDDistanceConstraintSystem:
    run_in_pause = False
    def update(self, world, dt):
        constraint_entities = world.query([DistanceConstraintComponent])
        epsilon = 1e-9 # Small value to avoid division by zero

        for entity_id in constraint_entities:
            constraint = world.get_component(entity_id, DistanceConstraintComponent)

            entity_a = constraint.entityA
            entity_b = constraint.entityB

            pos_a_comp = world.get_component(entity_a, PositionComponent)
            pos_b_comp = world.get_component(entity_b, PositionComponent)

            if not pos_a_comp or not pos_b_comp:
                continue

            p_a = pos_a_comp.pos
            p_b = pos_b_comp.pos

            mass_a_comp = world.get_component(entity_a, MassComponent)
            inv_mass_a = 1.0 / mass_a_comp.mass if mass_a_comp and mass_a_comp.mass > 0 else 0.0
            moi_a_comp = world.get_component(entity_a, MomentOfInertiaComponent)
            inv_inertia_a = moi_a_comp.inv_inertia if moi_a_comp else 0.0

            mass_b_comp = world.get_component(entity_b, MassComponent)
            inv_mass_b = 1.0 / mass_b_comp.mass if mass_b_comp and mass_b_comp.mass > 0 else 0.0
            moi_b_comp = world.get_component(entity_b, MomentOfInertiaComponent)
            inv_inertia_b = moi_b_comp.inv_inertia if moi_b_comp else 0.0

            if inv_mass_a + inv_mass_b + inv_inertia_a + inv_inertia_b <= epsilon:
                continue

            diff = p_b - p_a
            current_length = np.linalg.norm(diff)
            if current_length <= epsilon:
                continue

            direction = diff / current_length

            C = current_length - constraint.rest_length

            alpha_tilde = constraint.compliance / (dt * dt)

            w_sum = inv_mass_a + inv_mass_b

            denominator = w_sum + alpha_tilde
            if denominator <= epsilon:
                continue

            delta_lambda = (-C - alpha_tilde * constraint.lambda_val) / denominator
            constraint.lambda_val += delta_lambda

            correction = direction * delta_lambda

            if inv_mass_a > 0:
                dp_a = correction * -inv_mass_a
                p_a += dp_a
            if inv_mass_b > 0:
                dp_b = correction * inv_mass_b
                p_b += dp_b

class MovementSystem:
    run_in_pause = False
    def update(self, world, dt):
        grabbed = world.get_resource('grabbedBall')
        linear_entities = world.query([PositionComponent, VelocityComponent])
        for entity_id in linear_entities:
            if entity_id == grabbed:
                continue
            pos_comp = world.get_component(entity_id, PositionComponent)
            vel_comp = world.get_component(entity_id, VelocityComponent)
            pos_comp.pos += vel_comp.vel * dt

class PrevFinalPosSystem:
    run_in_pause = False
    def update(self, world, dt):
        entities = world.query([PositionComponent, PrevFinalPosComponent])
        for entity_id in entities:
            pos_component = world.get_component(entity_id, PositionComponent)
            prev_final_pos_component = world.get_component(entity_id, PrevFinalPosComponent)
            prev_final_pos_component.pos[:] = pos_component.pos

class AngularMovementSystem:
    run_in_pause = False
    def update(self, world, dt):
        entities = world.query([OrientationComponent, AngularVelocityComponent])
        for entity_id in entities:
            orientation = world.get_component(entity_id, OrientationComponent)
            angular_vel = world.get_component(entity_id, AngularVelocityComponent)
            orientation.angle += angular_vel.angular_velocity * dt

File: src/js/cable_joints/commonSystems.js
import Vector2 from './vector2.js';

import {
  PositionComponent,
  VelocityComponent,
  GravityAffectedComponent,
  MassComponent,
  PrevFinalPosComponent,
  OrientationComponent,
  MomentOfInertiaComponent,
  AngularVelocityComponent,
  DistanceConstraintComponent,
  PrevFinalOrientationComponent,
  RigidGroupComponent
} from './ecs.js';

export class PrevFinalOrientationSystem {
    runInPause = false;
    update(world, dt) {
        const entities = world.query([OrientationComponent, PrevFinalOrientationComponent]);
        for (const entityId of entities) {
            const orientationComp = world.getComponent(entityId, OrientationComponent);
            const prevFinalOrientationComp = world.getComponent(entityId, PrevFinalOrientationComponent);
            prevFinalOrientationComp.angle = orientationComp.angle;
        }
    }
}

export class PBDAngularVelocityUpdateSystem {
  runInPause = false;

  _normalizeAngle(angle) { // Helper to find the shortest angle difference
    while (angle > Math.PI) angle -= 2 * Math.PI;
    while (angle < -Math.PI) angle += 2 * Math.PI;
    return angle;
  }

  update(world, dt) {
    const grabbed = world.getResource('grabbedBall');
    const entities = world.query([
      OrientationComponent,
      AngularVelocityComponent,
      PrevFinalOrientationComponent,
      MomentOfInertiaComponent
    ]);
    const epsilon = 1e-9;

    if (dt <= epsilon) return;

    for (const entityId of entities) {
      if (entityId === grabbed) continue;

      const moiComp = world.getComponent(entityId, MomentOfInertiaComponent);
      if (moiComp && moiComp.invInertia <= 0) continue; // Skip static or non-rotational objects

      const orientationComp = world.getComponent(entityId, OrientationComponent);
      const angularVelComp = world.getComponent(entityId, AngularVelocityComponent);
      const prevFinalOrientationComp = world.getComponent(entityId, PrevFinalOrientationComponent);

      const currentAngle = orientationComp.angle;
      const prevAngle = prevFinalOrientationComp.angle;

      let deltaAngle = this._normalizeAngle(currentAngle - prevAngle);

      angularVelComp.angularVelocity = deltaAngle / dt;
    }
  }
}

export class PBDVelocityUpdateSystem {
  runInPause = false;
  update(world, dt) {
    const grabbed = world.getResource('grabbedBall');
    const entities = world.query([PositionComponent, VelocityComponent, PrevFinalPosComponent, MassComponent]);
    const epsilon = 1e-9;

    if (dt <= epsilon) return; // Avoid division by zero or very small dt

    for (const entityId of entities) {
      if (entityId === grabbed) continue; // Don't update velocity of grabbed ball this way

      const massComp = world.getComponent(entityId, MassComponent);
      // Only update velocities for dynamic objects (mass > 0)
      if (massComp && massComp.mass <= 0) continue;


      const posComp = world.getComponent(entityId, PositionComponent);
      const velComp = world.getComponent(entityId, VelocityComponent);
      const prevFinalPosComp = world.getComponent(entityId, PrevFinalPosComponent);

      // v = (x_new - x_old) / dt
      velComp.vel.subtractVectors(posComp.pos, prevFinalPosComp.pos).scale(1.0 / dt);
    }
  }
}


export class GravitySystem {
  runInPause = false;
  update(world, dt) {
    const grabbed = world.getResource('grabbedBall');
    const gravity = world.getResource('gravity');
    if (!gravity) return;

    const entities = world.query([VelocityComponent, GravityAffectedComponent]);
    for (const entityId of entities) {
      if (entityId === grabbed) continue;
      const velComp = world.getComponent(entityId, VelocityComponent);
      velComp.vel.add(gravity, dt);
    }
  }
}

export class XPBDDistanceConstraintSystem {
  runInPause = false;
  update(world, dt) {
    const constraintEntities = world.query([DistanceConstraintComponent]);
    const epsilon = 1e-9; // Small value to avoid division by zero

    for (const entityId of constraintEntities) {
      const constraint = world.getComponent(entityId, DistanceConstraintComponent);

      const entityA = constraint.entityA;
      const entityB = constraint.entityB;

      const posAComp = world.getComponent(entityA, PositionComponent);
      const posBComp = world.getComponent(entityB, PositionComponent);

      if (!posAComp || !posBComp) continue;

      const pA = posAComp.pos;
      const pB = posBComp.pos;

      const massAComp = world.getComponent(entityA, MassComponent);
      const invMassA = (massAComp && massAComp.mass > 0) ? 1.0 / massAComp.mass : 0.0;
      const moiAComp = world.getComponent(entityA, MomentOfInertiaComponent);
      const invInertiaA = moiAComp ? moiAComp.invInertia : 0.0;

      const massBComp = world.getComponent(entityB, MassComponent);
      const invMassB = (massBComp && massBComp.mass > 0) ? 1.0 / massBComp.mass : 0.0;
      const moiBComp = world.getComponent(entityB, MomentOfInertiaComponent);
      const invInertiaB = moiBComp ? moiBComp.invInertia : 0.0;

      // If both entities are effectively immovable, skip
      if (invMassA + invMassB + invInertiaA + invInertiaB <= epsilon) {
        continue;
      }

      const diff = new Vector2().subtractVectors(pB, pA);
      const currentLength = diff.length();
      if (currentLength <= epsilon) continue;

      const dir = diff.clone().scale(1.0 / currentLength); // Normalized direction from A to B

      const C = currentLength - constraint.restLength; // Constraint violation

      // XPBD compliance term (alpha_tilde in the paper)
      const alpha_tilde = constraint.compliance / (dt * dt);

      // Generalized inverse mass for the constraint
      // For distance constraints, the gradient w.r.t position is 'dir' or '-dir'
      // For rotation, it's more complex if attachment points are offset, but here we assume CoM constraint
      // For simplicity, assuming constraint acts on CoM directly (no rotational coupling here)
      // More general XPBD would include r x n terms for rotational parts.
      // This simplified version matches standard PBD distance constraint generalized mass.
      let w_sum = invMassA + invMassB; // Simplified: only translational

      // Denominator for delta_lambda
      const denominator = w_sum + alpha_tilde;
      if (denominator <= epsilon) continue;

      const delta_lambda = (-C - alpha_tilde * constraint.lambda) / denominator;
      constraint.lambda += delta_lambda;

      const correction = dir.clone().scale(delta_lambda);

      // Apply positional corrections
      if (invMassA > 0) {
        const dpA = correction.clone().scale(-invMassA);
        pA.add(dpA);
        const velAComp = world.getComponent(entityA, VelocityComponent);
        if (velAComp && dt > epsilon) {
            // velAComp.vel.add(dpA, 1.0 / dt); // This would be for PBD velocity update
        }
      }
      if (invMassB > 0) {
        const dpB = correction.clone().scale(invMassB);
        pB.add(dpB);
        const velBComp = world.getComponent(entityB, VelocityComponent);
        if (velBComp && dt > epsilon) {
            // velBComp.vel.add(dpB, 1.0 / dt); // This would be for PBD velocity update
        }
      }
    }
  }
}

// Enforce rigid motion of grouped bodies via 2D shape matching (best-fit rigid transform).
// References: Müller et al. 2005 (shape matching), PBDBodies (rigid body constraints),
// but implemented as a single-shot Jacobi-style correction per group to avoid order bias.
export class RigidGroupSystem {
  runInPause = false;

  _computeCOM(world, members) {
    let sumMass = 0.0;
    const com = new Vector2(0, 0);
    for (const id of members) {
      const pos = world.getComponent(id, PositionComponent)?.pos;
      const m = world.getComponent(id, MassComponent)?.mass ?? 0.0;
      if (!pos || !(m > 0)) continue;
      com.add(pos.clone().scale(m));
      sumMass += m;
    }
    if (sumMass > 0) com.scale(1.0 / sumMass);
    return { com, sumMass };
  }

  update(world, dt) {
    const groupEntities = world.query([RigidGroupComponent]);
    if (!groupEntities || groupEntities.length === 0) return;

    for (const gid of groupEntities) {
      const group = world.getComponent(gid, RigidGroupComponent);
      const members = group.members || [];
      if (members.length < 2) continue;

      // Initialize rest offsets once from current configuration
      if (!group.restLocal) {
        const { com } = this._computeCOM(world, members);
        group.restLocal = members.map((id) => {
          const pos = world.getComponent(id, PositionComponent)?.pos;
          return pos ? pos.clone().subtract(com) : new Vector2(0, 0);
        });
      }

      // Current COM and relative positions
      const { com, sumMass } = this._computeCOM(world, members);
      if (!(sumMass > 0)) continue;

      // Compute best-fit rotation (2D) mapping restLocal -> currentRel using weighted Procrustes
      let Sx = 0.0; // sum m (r.x*p.x + r.y*p.y)
      let Sy = 0.0; // sum m (r.x*p.y - r.y*p.x)
      for (let i = 0; i < members.length; i++) {
        const id = members[i];
        const pos = world.getComponent(id, PositionComponent)?.pos;
        const m = world.getComponent(id, MassComponent)?.mass ?? 0.0;
        if (!pos || !(m > 0)) continue;
        const p = pos.clone().subtract(com);
        const r = group.restLocal[i] || new Vector2(0, 0);
        Sx += m * (r.x * p.x + r.y * p.y);
        Sy += m * (r.x * p.y - r.y * p.x);
      }

      const angle = Math.atan2(Sy, Sx);
      const c = Math.cos(angle);
      const s = Math.sin(angle);

      // Target positions under best-fit rigid transform
      const stiffness = Math.max(0, Math.min(1, group.stiffness ?? 1.0));
      for (let i = 0; i < members.length; i++) {
        const id = members[i];
        const posComp = world.getComponent(id, PositionComponent);
        if (!posComp) continue;
        const r = group.restLocal[i] || new Vector2(0, 0);
        const qx = c * r.x - s * r.y + com.x;
        const qy = s * r.x + c * r.y + com.y;
        const dp = new Vector2(qx - posComp.pos.x, qy - posComp.pos.y).scale(stiffness);
        posComp.pos.add(dp);
      }

      // Apply group rotation to each member's local orientation
      let deltaAngle = angle - (group.prevAngle || 0.0);
      // normalize to [-pi, pi] for shortest rotation
      while (deltaAngle > Math.PI) deltaAngle -= 2 * Math.PI;
      while (deltaAngle < -Math.PI) deltaAngle += 2 * Math.PI;
      if (Math.abs(deltaAngle) > 0) {
        for (let i = 0; i < members.length; i++) {
          const id = members[i];
          const o = world.getComponent(id, OrientationComponent);
          if (o) {
            o.angle += deltaAngle;
          }
        }
      }

      // Optional debug output if enabled
      try {
        if (world.getResource && world.getResource('debugAngles')) {
          const first = members[0];
          const o0 = world.getComponent(first, OrientationComponent);
          console.log('[RigidGroupSystem]', { angle, deltaAngle, o0: o0 ? o0.angle : null });
        }
      } catch(_) {}

      group.prevAngle = angle;
    }
  }
}

export class MovementSystem {
  runInPause = false;
  update(world, dt) {
    const grabbed = world.getResource('grabbedBall');
    // Update linear position
    const linearEntities = world.query([PositionComponent, VelocityComponent]);
    for (const entityId of linearEntities) {
      if (entityId === grabbed) continue;
      const posComp = world.getComponent(entityId, PositionComponent);
      const velComp = world.getComponent(entityId, VelocityComponent);
      posComp.pos.add(velComp.vel, dt);
    }
  }
}

export class PrevFinalPosSystem {
    runInPause = false;
    update(world, dt) {
        const entities = world.query([PositionComponent, PrevFinalPosComponent]);
        for (const entityId of entities) {
            const posComponent = world.getComponent(entityId, PositionComponent);
            const prevFinalPosComponent = world.getComponent(entityId,PrevFinalPosComponent);
            prevFinalPosComponent.pos.set(posComponent.pos);
        }
    }
}


export class AngularMovementSystem {
    runInPause = false;
    update(world, dt) {
        const entities = world.query([OrientationComponent, AngularVelocityComponent]);
        for (const entityId of entities) {
            const orientation = world.getComponent(entityId, OrientationComponent);
            const angularVel = world.getComponent(entityId, AngularVelocityComponent);

            orientation.angle += angularVel.angularVelocity * dt;
        }
    }
}
